File: android/app/src/main/kotlin/pt/ualg/miaugenda/data/remote/RequestApi.kt
package pt.ualg.miaugenda.data.remote

import pt.ualg.miaugenda.data.model.TimeOffRequest
import pt.ualg.miaugenda.data.model.ShiftSwapRequest
import pt.ualg.miaugenda.data.model.CreateTimeOffRequestBody
import pt.ualg.miaugenda.data.model.CreateShiftSwapRequestBody
import retrofit2.Response
import retrofit2.http.Body
import retrofit2.http.DELETE
import retrofit2.http.GET
import retrofit2.http.PATCH
import retrofit2.http.POST
import retrofit2.http.Path

interface RequestApi {

    @GET("/time-off-requests")
    suspend fun getTimeOffRequests(): Response<List<TimeOffRequest>>

    @POST("/time-off-requests")
    suspend fun createTimeOffRequest(
        @Body body: CreateTimeOffRequestBody
    ): Response<TimeOffRequest>

    @PATCH("/time-off-requests/{id}/status")
    suspend fun updateTimeOffRequestStatus(
        @Path("id") id: Int,
        @Body body: Map<String, String>
    ): Response<TimeOffRequest>

    @DELETE("/time-off-requests/{id}")
    suspend fun deleteTimeOffRequest(
        @Path("id") id: Int
    ): Response<Unit>

    @GET("/shift-swap-requests")
    suspend fun getShiftSwapRequests(): Response<List<ShiftSwapRequest>>

    @POST("/shift-swap-requests")
    suspend fun createShiftSwapRequest(
        @Body body: CreateShiftSwapRequestBody
    ): Response<ShiftSwapRequest>

    @PATCH("/shift-swap-requests/{id}/status")
    suspend fun updateShiftSwapRequestStatus(
        @Path("id") id: Int,
        @Body body: Map<String, String>
    ): Response<ShiftSwapRequest>

    @PATCH("/shift-swap-requests/{id}/target-response")
    suspend fun respondToSwapRequest(
        @Path("id") id: Int,
        @Body body: Map<String, Boolean>
    ): Response<ShiftSwapRequest>

    @DELETE("/shift-swap-requests/{id}")
    suspend fun deleteShiftSwapRequest(
        @Path("id") id: Int
    ): Response<Unit>
}

File: desktop/src/api/requestApi.ts
import { apiClient } from "./apiClient";
import type {
  CreateShiftSwapRequestBody,
  CreateTimeOffRequestBody,
  ShiftSwapRequest,
  TimeOffRequest,
  UpdateRequestStatusBody
} from "../models/request";

export async function getTimeOffRequests(params?: { userId?: number; from?: string; to?: string }): Promise<TimeOffRequest[]> {
  const response = await apiClient.get<TimeOffRequest[]>("/time-off-requests", { params });
  return response.data;
}

export async function createTimeOffRequest(
  body: CreateTimeOffRequestBody
): Promise<TimeOffRequest> {
  const response = await apiClient.post<TimeOffRequest>("/time-off-requests", body);
  return response.data;
}

export async function updateTimeOffRequestStatus(
  id: number,
  body: UpdateRequestStatusBody
): Promise<TimeOffRequest> {
  const response = await apiClient.patch<TimeOffRequest>(
    `/time-off-requests/${id}/status`,
    body
  );

  return response.data;
}

export async function getShiftSwapRequests(params?: { userId?: number; from?: string; to?: string }): Promise<ShiftSwapRequest[]> {
  const response = await apiClient.get<ShiftSwapRequest[]>("/shift-swap-requests", { params });
  return response.data;
}

export async function createShiftSwapRequest(
  body: CreateShiftSwapRequestBody
): Promise<ShiftSwapRequest> {
  const response = await apiClient.post<ShiftSwapRequest>("/shift-swap-requests", body);
  return response.data;
}

export async function updateShiftSwapRequestStatus(
  id: number,
  body: UpdateRequestStatusBody
): Promise<ShiftSwapRequest> {
  const response = await apiClient.patch<ShiftSwapRequest>(
    `/shift-swap-requests/${id}/status`,
    body
  );

  return response.data;
}

export async function respondToSwapRequest(
  id: number,
  accept: boolean
): Promise<ShiftSwapRequest> {
  const response = await apiClient.patch<ShiftSwapRequest>(
    `/shift-swap-requests/${id}/target-response`,
    { accept }
  );

  return response.data;
}